### PR TITLE
Rpi conn documentation

### PR DESCRIPTION
**Pull request by derins**
Endpoint for SNS connection created
- prints subscription confirmation url
- parses data when subscribed

RPi documentation written

### CODE REVIEW

File: web/web/api/v1/sns_read.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+from flask import Flask, request, Blueprint
+import json
+
+sns_read_api_bp =  Blueprint('sns_read_api_bp', __name__)
```
Comment (jongoncalves) [inline web/web/api/v1/sns_read.py:4]: why `read` in the name? Let's just make it `sns_api_bp`

File: web/web/api/v1/sns_read.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+
+@sns_read_api_bp.route('/sns_read', methods = ['GET', 'POST'])
```
Comment (jongoncalves) [inline web/web/api/v1/sns_read.py:7]: does AWS really POST and GET to this endpoint? Seems weird to support both

File: web/web/api/v1/sns_read.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+def sns():
```
Comment (jongoncalves) [inline web/web/api/v1/sns_read.py:8]: Should this be called `confirm_subscription` instead?

File: web/web/api/v1/sns_read.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    # AWS sends JSON with text/plain mimetype
+    try:
+        js = json.loads(request.data)
+    except:
+        pass
```
Comment (jongoncalves) [inline web/web/api/v1/sns_read.py:13]: if the parsing throws an exception and you pass the handling, line 17 and 19 would definitely blow up if reached. Either return early with an error or let it crash.

File: web/web/api/v1/sns_read.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    hdr = request.headers.get('x-amz-sns-message-type')
+    # subscribe to the SNS topic
+    if hdr == 'Notification':
+        parseRealPower(js['Message'])
+    else if hdr == 'SubscriptionConfirmation' and 'SubscribeURL' in js:
+        return js['SubscribeURL']
+    return 'OK\n'
+
+
+def parseRealPower(message):
```
Comment (jongoncalves) [inline web/web/api/v1/sns_read.py:23]: snake case

File: web/web/api/v1/sns_read.py (hunk 3 above)
Comment (jongoncalves) [reply web/web/api/v1/sns_read.py:8]: I realized this receives the confirmation and the message from the devices. Maybe we can call this `subscriber_information` or `listener`, or something of the sorts

File: web/web/api/v1/sns_read.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    ''' converts byte wtring to dict and retrieves a sunnyboy_inverter.calc_ac_power_kw
+        data and timestamp as a dict
+    '''
+    payload_str = message.payload.decode("UTF-8")
```
Comment (jongoncalves) [inline web/web/api/v1/sns_read.py:27]: weird that this isn't json.....
Just making a note here for posterity that we'll need to update this method to insert the data in the the DB

File: web/web/api/v1/sns_read.py (hunk 1 above)
Comment (derins) [reply web/web/api/v1/sns_read.py]: done

File: web/web/api/v1/sns_read.py (hunk 2 above)
Comment (derins) [reply web/web/api/v1/sns_read.py]: You're right, converted to post only

File: web/web/api/v1/sns_read.py (hunk 3 above)
Comment (derins) [reply web/web/api/v1/sns_read.py]: done

File: web/web/api/v1/sns_read.py (hunk 4 above)
Comment (derins) [reply web/web/api/v1/sns_read.py]: Not sure if I'm handling this the way you want it, lmk if there needs to be a change - I'm just returning an error message now

File: web/web/api/v1/sns_read.py (hunk 5 above)
Comment (derins) [reply web/web/api/v1/sns_read.py]: fixed

File: web/web/api/v1/sns_read.py (hunk 6 above)
Comment (derins) [reply web/web/api/v1/sns_read.py]: Do you need me to pin this as an issue?

**Comment by jongoncalves** [comment]
You seem to have deleted the sns file - was that intentional or did you forget to add a new file?

**Comment by derins** [comment]
@jongoncalves added the new file - my bad

File: web/web/api/v1/sns.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+        js = json.loads(request.data)
+    except:
+        return 'An error occuered'
+    hdr = request.headers.get('x-amz-sns-message-type')
+    # subscribe to the SNS topic
+    if hdr == 'Notification':
+        parse_real_power(js['Message'])
+    elif hdr == 'SubscriptionConfirmation' and 'SubscribeURL' in js:
+        return js['SubscribeURL']
+    return 'OK'
```
Comment (jongoncalves) [inline web/web/api/v1/sns.py:20]: This still seems pretty weird to me, a couple of things:
- Do we actually need to return anything in any of the code paths?
- According to SNS docs, to confirm subscription we just need to visit the URL. Does `return`ing actually do this for us or should we issue a request to that URL?
- Also, we need to be explicit about which error we catching. Don't just catch all exceptions.
- If we don't need to return anything, let's just log the errors/success

File: web/web/api/v1/sns.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+
+def parse_real_power(message):
```
Comment (jongoncalves) [inline web/web/api/v1/sns.py:23]: Can you share an example of what this message looks like? From my understanding this should be json already....I'm wondering if it's double encoded out of curiosity.

Comment (derins) [reply web/web/api/v1/sns.py:23]: b'{"DeviceID": "TEC_GISMO", "DeviceInformation": [{"sunnyboy_inverter.power_l1": {"timestamp": 1606332463830, "value": 0.0}, "sunnyboy_inverter.reactive_power_l2": {"timestamp": 1606332463830, "value": 0.0}, "sunnyboy_inverter.calc_ac_power_kw": {"timestamp": 1606332463830, "value": 0.0}, "sunnyboy_inverter.reactive_power_l1": {"timestamp": 1606332463830, "value": 0.0}, "sunnyboy_inverter.dc_voltage_input_1": {"timestamp": 1606332463830, "value": 394.07}, "sunnyboy_inverter.dc_voltage_input_2": {"timestamp": 1606332463830, "value": 395.38}, "sunnyboy_inverter.grid_voltage_phase_l2": {"timestamp": 1606332463830, "value": 117.26}, "sunnyboy_inverter.apparent_power_l1": {"timestamp": 1606332463830, "value": 0.0}, "sunnyboy_inverter.apparent_power_l2": {"timestamp": 1606332463830, "value": 0.0}, "sunnyboy_inverter.grid_voltage_phase_l1": {"timestamp": 1606332463830, "value": 117.14}, "sunnyboy_inverter.grid_frequency": {"timestamp": 1606332463830, "value": 60.02}, "sunnyboy_inverter.power_l2": {"timestamp": 1606332463830, "value": 0.0}, "sunnyboy_inverter.dc_current_input_1": {"timestamp": 1606332463830, "value": 0.0}, "sunnyboy_inverter.dc_current_input_2": {"timestamp": 1606332463830, "value": 0.0}, "sunnyboy_inverter.dc_power_input_2": {"timestamp": 1606332463830, "value": 0.0}, "sunnyboy_inverter.dc_power_input_1": {"timestamp": 1606332463830, "value": 0.0}, "sunnyboy_inverter.grid_current_phase_l1": {"timestamp": 1606332463830, "value": 0.0}, "sunnyboy_inverter.grid_current": {"timestamp": 1606332463830, "value": 0.0}, "sunnyboy_inverter.gridms_phv_a2b": {"timestamp": 1606332463830, "value": 234.4}, "sunnyboy_inverter.grid_current_phase_l2": {"timestamp": 1606332463830, "value": 0.0}}]}'

File: web/web/api/v1/sns.py (hunk 7 above)
Comment (derins) [reply web/web/api/v1/sns.py]: The endpoint errors out without a return at the end, I made changes for the second and third bullet point. The function still seems to require me to return something - I also added the logger.

**Review by jongoncalves — APPROVED** [review]

**Review by avpeery — APPROVED** [review]

**Pull request merged**